tool/tctl/common/terraform_command.go

```diff
@@ -0,0 +1,380 @@
+	id, err := c.useBotToObtainIdentity(ctx, addr, tokenName, client)
+	if err != nil {
+		return trace.Wrap(err, "obtaining identity")
+	}
+
+	envVars, err := identityToTerraformEnvVars(addr.String(), id)
+	if err != nil {
+		return trace.Wrap(err, "exporting identity into environment variables")
+	}
+
+	// Export environment variables
+	c.showProgress("Certificates obtained, you can now use Terraform in this terminal 🚀")
+	for env, value := range envVars {
+		_, _ = fmt.Fprintf(c.envOutput, "export %s=%q\n", env, value)
+	}
+	fmt.Println("# You must invoke this command in an eval: eval $(tctl terraform-helper)")
+	return nil
+}
+
+// createTransientBotAndToken creates a Bot resource and a secret Token.
+// The token is single use (secret tokens are consumed on MachineID join)
+// and the bot expires after the given TTL.
+func (c *TerraformCommand) createTransientBotAndToken(ctx context.Context, client *authclient.Client, roleName string) (string, error) {
+	// Create token and bot name
+	suffix, err := utils.CryptoRandomHex(4)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	botName := c.resourcePrefix + suffix
+	c.showProgress(fmt.Sprintf("Creating temporary bot %q and its token", botName))
+
+	roles := []string{roleName}
+	var token types.ProvisionToken
+
+	// Generate a token
+	tokenName, err := utils.CryptoRandomHex(defaults.TokenLenBytes)
+	if err != nil {
+		return "", trace.Wrap(err, "generating random token")
+	}
+	tokenSpec := types.ProvisionTokenSpecV2{
+		Roles:      types.SystemRoles{types.RoleBot},
+		JoinMethod: types.JoinMethodToken,
+		BotName:    botName,
+	}
+	// Token should be consumed on bot join in a few seconds. If the bot fails to join for any reason,
+	// the token should not outlive the bot.
+	token, err = types.NewProvisionTokenFromSpec(tokenName, time.Now().Add(c.botTTL), tokenSpec)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	if err := client.UpsertToken(ctx, token); err != nil {
+		return "", trace.Wrap(err, "upserting token")
+	}
+
+	// Create bot
+	bot := &machineidv1pb.Bot{
+		Metadata: &headerv1.Metadata{
+			Name:    botName,
+			Expires: timestamppb.New(time.Now().Add(c.botTTL)),
+		},
+		Spec: &machineidv1pb.BotSpec{
+			Roles: roles,
+		},
+	}
+
+	bot, err = client.BotServiceClient().CreateBot(ctx, &machineidv1pb.CreateBotRequest{
+		Bot: bot,
+	})
+	if err != nil {
+		return "", trace.Wrap(err, "creating bot")
+	}
+	return tokenName, nil
+}
+
+// roleClient describes the minimal set of operations that the helper uses to
+// create the Terraform provider role.
+type roleClient interface {
+	UpsertRole(context.Context, types.Role) (types.Role, error)
+	GetRole(context.Context, string) (types.Role, error)
+}
+
+// createRoleIfNeeded upserts the Terraform role, or checks if the role exists.
+// Returns the Terraform role name.
+func (c *TerraformCommand) createRoleIfNeeded(ctx context.Context, client roleClient) (string, error) {
+	log := slog.Default()
```

Comment:
Let's create this way up and re-use it here

Reply:
There are a lot of commands in this package and I'm not super comfortable with defining a package-wide variable called "log", this will lead to conflicts and potential bugs. The best I can do is pass the logger as a function argument.

Reply:
Sorry, I should've been clearer.
Didn't mean to suggest as a package-wide variable but a field in `TerraformCommand` struct.